### PR TITLE
[studio-ui] Content Tree Panel #3809

### PR DESCRIPTION
**Pull request by aachavarria**
https://github.com/craftercms/craftercms/issues/3809

Adding support for Components/Pages on Root Lv;

### CODE REVIEW

File: ui/guest/src/classes/ElementRegistry.ts [hunk 1]
```diff
@@ -155,6 +160,7 @@ export function deregister(id: string | number): ElementRecord {
     const { iceIds } = record;
     iceIds.forEach((iceId) => iceRegistry.deregister(iceId));
     delete db[id];
+    //TODO: delete from registry;
```
Comment (rart) [inline ui/guest/src/classes/ElementRegistry.ts]: 👆🏻

File: ui/guest/src/classes/ElementRegistry.ts [hunk 2]
```diff
@@ -186,11 +192,38 @@ export function getRect(id: number): DOMRect {
...
 export function fromICEId(iceId: number): RegistryEntry {
-  return registry[iceId];
-  // return Object.values(db).find(({ iceIds }) => {
-  //   return iceIds.includes(iceId);
-  // });
+  const record = db[registry[iceId]?.[0]];
+  return fromRecord(record, iceId);
+}
+
+export function getRegistriesFromICEId(iceId: number): RegistryEntry[] {
```
Comment (rart) [inline ui/guest/src/classes/ElementRegistry.ts]: getRecordsFromIceId

File: ui/guest/src/classes/ElementRegistry.ts [hunk 3]
```diff
@@ -186,11 +192,38 @@ export function getRect(id: number): DOMRect {
...
+  const recordsIds = registry[iceId];
+  const registries = [];
```
Comment (rart) [inline ui/guest/src/classes/ElementRegistry.ts]: records

File: ui/guest/src/constants.ts [hunk 4]
```diff
@@ -59,6 +59,7 @@ export const CLEAR_HIGHLIGHTED_RECEPTACLES = 'CLEAR_HIGHLIGHTED_RECEPTACLES';
 export const CONTENT_TREE_FIELD_SELECTED = 'CONTENT_TREE_FIELD_SELECTED';
 export const CLEAR_CONTENT_TREE_FIELD_SELECTED = 'CLEAR_CONTENT_TREE_FIELD_SELECTED';
 export const CHILDREN_MAP_UPDATE = 'CHILDREN_MAP_UPDATE';
+export const CONTENT_TREE_FIELD_CHANGE_ELEMENT = 'CONTENT_TREE_FIELD_CHANGE_ELEMENT';
```
Comment (rart) [inline ui/guest/src/constants.ts]: CONTENT_TREE_SWITCH_FIELD

File: ui/guest/src/env.ts [hunk 5]
```diff
@@ -32,6 +32,7 @@ declare global {
       'craftercms-asset-uploader-mask': CrafterCMSCustomElementProps;
       'craftercms-snackbar': CrafterCMSCustomElementProps;
       'craftercms-dragged-element': CrafterCMSCustomElementProps;
+      'craftercms-element-selector': CrafterCMSCustomElementProps;
```
Comment (rart) [inline ui/guest/src/env.ts]: craftercms-field-instance-switcher

File: ui/guest/src/react/ElementSelector.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from 'react';
+
+interface ElementSelector {
```
Comment (rart) [inline ui/guest/src/react/ElementSelector.tsx]: FieldInstanceSwitcherProps

File: ui/guest/src/react/ElementSelector.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  currentElement: number;
+  registryEntryIds: number[];
+  onNext(): void;
+  onPrev(): void;
+}
+
+export default function ElementSelector(props: ElementSelector) {
+  const {
+    onNext,
+    onPrev,
```
Comment (rart) [inline ui/guest/src/react/ElementSelector.tsx]: FieldInstanceSwitcher

File: ui/guest/src/react/ElementSelector.tsx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    registryEntryIds,
+    currentElement
+  } = props;
+  return (
+    <craftercms-element-selector>
```
Comment (rart) [inline ui/guest/src/react/ElementSelector.tsx]: craftercms-field-instance-switcher

File: ui/guest/src/react/Guest.tsx [hunk 9]
```diff
@@ -40,6 +40,7 @@ import {
   COMPONENT_DRAG_STARTED,
   COMPONENT_INSTANCE_DRAG_ENDED,
   COMPONENT_INSTANCE_DRAG_STARTED,
+  CONTENT_TREE_FIELD_CHANGE_ELEMENT,
```
Comment (rart) [inline ui/guest/src/react/Guest.tsx]: CONTENT_TREE_SWITCH_FIELD_INSTANCE

File: ui/guest/src/store/models/GuestStore.ts [hunk 10]
```diff
@@ -57,6 +57,11 @@ export interface GuestState {
   uploading: LookupTable;
   models: LookupTable<ContentInstance>;
   contentTypes: LookupTable<ContentType>;
+  elementSelector?: {
```
Comment (rart) [inline ui/guest/src/store/models/GuestStore.ts]: fieldSwitcher

File: ui/app/src/modules/Preview/Tools/ContentTree.tsx [hunk 11]
```diff
@@ -668,16 +730,33 @@ function ContentTreeUI(props: ContentTreeUI) {
...
-  const data = resource.read();
-  const { breadcrumbs, nodeLookup, selected } = data;
+  resource.read();
+
+  let node: any = null;
+
+  if (selected === 'root') {
+    node = {
+      id: 'root',
+      name: 'On this page...',
```
Comment (rart) [inline ui/app/src/modules/Preview/Tools/ContentTree.tsx]: Translation

File: ui/guest/src/classes/ElementRegistry.ts [hunk 12]
```diff
@@ -290,14 +330,14 @@ export function getDragContextFromReceptacles(
...
-export function getElementFromICEProps(modelId: string, fieldId: string, index: string | number): JQuery<Element> {
+export function getElementFromICEProps(modelId: string, fieldId: string, index: string | number): Element {
   const recordId = iceRegistry.exists({
     modelId: modelId,
     fieldId: fieldId,
     index: index
   });
 
-  return (recordId === -1 || !fromICEId(recordId)) ? null : $(fromICEId(recordId).element);
+  return (recordId === -1 || !fromICEId(recordId)) ? null : fromICEId(recordId).element;
```
Comment (rart) [inline ui/guest/src/classes/ElementRegistry.ts]: Avoid calling fromICEId twice.

File: ui/guest/src/classes/ElementRegistry.ts [hunk 13]
```diff
@@ -186,11 +199,38 @@ export function getRect(id: number): DOMRect {
   return get(id).element.getBoundingClientRect();
 }
 
+export function fromRecord(record: ElementRecord, iceId: number): RegistryEntry {
+  if (!record) {
+    return null;
+  }
+  const { id, element, label, modelId, index, fieldId: fieldIds } = record;
+  return { id, element, modelId, index, label, fieldId: fieldIds[0], iceId };
```
Comment (rart) [inline ui/guest/src/classes/ElementRegistry.ts]: This is fixed to always read the first field ID, disregarding the actual field ID on the desired ice record.

File: ui/guest/src/classes/ElementRegistry.ts [hunk 14]
```diff
@@ -186,11 +199,38 @@ export function getRect(id: number): DOMRect {
   return get(id).element.getBoundingClientRect();
 }
 
+export function fromRecord(record: ElementRecord, iceId: number): RegistryEntry {
```
Comment (rart) [inline ui/guest/src/classes/ElementRegistry.ts]: Rename to `createIntermediateElementRecord`

File: ui/guest/src/classes/ElementRegistry.ts (hunk 1 above)
Comment (rart) [reply ui/guest/src/classes/ElementRegistry.ts]: 👆🏻

File: ui/guest/src/classes/ElementRegistry.ts (hunk 13 above)
Comment (aachavarria) [reply ui/guest/src/classes/ElementRegistry.ts:207]: yep, and that is ok, there are another method to read all the ID's

**Review by rart — APPROVED** [review]

**Pull request merged**